### ISSUE TITLE
Bug: [no-unnecessary-condition] false positive when assigning in a forEach loop

### ISSUE THREAD

**Issue by ChristophP** [typescript-eslint/typescript-eslint#7540]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.1-rc&fileType=.tsx&code=GYVwdgxgLglg9mABBAFgUwgawBQAcBOcuAzgJSIDeAUIogDZpSIBuAhnYgLyLDvFoBuKjUQBtAEStExGGADmDRDChoAtuIC6AOmBx8AUVaps2cpwB8lEbRjBEeQiS25GafOWq0vLdl0RR8EEFrRABfEVDSIRFbezY6DxCIBGI4Bi06ODlscQAVdGQEABNleCQYYn9AtHEoiKpwoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
function check(props) {
  let val = false;

  ["a single item"].forEach(() => {
    if (props.peter) {
      val = true;
    }
  });

  if (val) {
    console.log("The condition is true");
  }
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect the rule not to assert that the final condition in the snippet is unnecessary and always falsy. Since the compiler does not know the type of `props` it can not know whether the condition will be true or false.

### Actual Result

Actually the compiler says the condition is unnecessy which is a false positive in my case as far as I can tell.

**Comment by bradzacher** [comment]
https://typescript-eslint.io/rules/no-unnecessary-condition#when-not-to-use-it

Please read the rule docs

**Comment by ChristophP** [comment]
Got it, thanks for the link.
Sorry for the false alarm.
